### PR TITLE
DAOS-9555 control: Detect UCX fabric interfaces

### PR DESCRIPTION
**Pull request by kjacque**
This patch enables detection of UCX fabric interfaces. Either
libfabric, UCX, or both may be installed at runtime. Both
libfabric and UCX are required at build time in order to build
with UCX support.

- Add dynamic UCX bindings and new UCX FabricInterfaceProvider.
- Add the UCX provider to the list of default
  FabricInterfaceProviders, if building with UCX.
- Fail early if no fabric provider is installed.
- Translate UCX transport strings to those used by the server.
- Refactor FabricInterface to better represent the distinction
  between name as reported by the fabric library, name as it
  appears in the hardware topology, and its associated net
  device. UCX reports a different device name than what appears
  in the topology.
- Conditionally build ucx support with scons flag UCX=1. Also
  updated Mercury build to use this flag instead of detecting
  the library at build time.
- Make the UCX OFI_DOMAIN name a combination of all device names.
  Not sure why this is required, but it is needed for multi-engine
  support.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Comment by kjacque** [comment]
@jgmoore-or if you have a chance, can you please verify that the latest and greatest version of this PR still works for you with UCX?

File: src/control/SConscript [hunk 1]
```diff
@@ -38,11 +38,20 @@ def is_firmware_mgmt_build(benv):
...
+
 def get_build_tags(benv):
     "Get custom go build tags."
-    if not is_firmware_mgmt_build(benv):
+    tags = ""
+    if is_firmware_mgmt_build(benv):
+        tags += "firmware "
+    if is_ucx_build(benv):
+        tags += "ucx "
+    if tags == "":
```
Comment (mjmac) [inline src/control/SConscript:52]: I'm pretty sure that `-tags` wants a comma-separated list of tags. In my catastrophic recovery branch, I made the following change:
```
-    if not is_firmware_mgmt_build(benv):
+    tags = []
+    if is_firmware_mgmt_build(benv):
+        tags.append("firmware")
+    if not is_release_build(benv):
+        tags.append("fault_injection")
+
+    if len(tags) == 0:
         return ""
-    return "-tags firmware"
+
+    return "-tags {}".format(','.join(tags))
```

**Review by mjmac — COMMENTED** [review]
Took a quick look, seems pretty straightforward. Will review in depth on Monday. Only thing I see that definitely needs to be fixed is the build tags constructor in the build goop.

**Review by tanabarr — APPROVED** [review]
I admit I don't understand all the details but from a go coding standpoint all looks good to me, great work!

File: src/control/SConscript [hunk 2]
```diff
@@ -44,14 +44,14 @@ def is_ucx_build(benv):
...
     "Get custom go build tags."
-    tags = ""
+    tags = []
     if is_firmware_mgmt_build(benv):
-        tags += "firmware "
+        tags.append("firmware")
     if is_ucx_build(benv):
-        tags += "ucx "
+        tags.append("ucx")
     if tags == "":
```
Comment (mjmac) [inline src/control/SConscript:52]: Oops, this test will always be false. I think you want the following:
```suggestion
    if len(tags) == 0:
```

Comment (kjacque) [reply src/control/SConscript:52]: Oops!

**Review by tanabarr — APPROVED** [review]

**Review by mjmac — DISMISSED** [review]

**Review by tanabarr — APPROVED** [review]

**Review by jgmoore-or — APPROVED** [review]

**Review by knard-intel — COMMENTED** [review]
OK for me.

**Review by knard-intel — APPROVED** [review]
Ok for me => better with the approvement ;)

**Pull request merged**